### PR TITLE
🤖 Set initial FTE TextLine properties and metrics

### PR DESCRIPTION
**Pull request by doctorpangloss**
## Description

This adds the first createTextLine behavior needed for Flash Text Engine TextLine support.

Flash Player returns a TextLine from TextBlock.createTextLine with observable slots for rawTextLength, textBlockBeginIndex, specifiedWidth, textBlock, previousLine, nextLine, and the TextBlock firstLine and lastLine links. It also exposes initial metric getters and TextLine.MAX_LINE_WIDTH.

This PR stores those TextLine fields, wires the line chain slots, and caches a TextLineLayout for initial textWidth, textHeight, ascent, and descent values. The metric path and TextFormat conversion are reported as stubs because later PRs add richer formatting and metric behavior.

The SWF tests cover the stored TextLine fields, TextBlock links, line chaining, default ElementFormat values, stub metric traces, and MAX_LINE_WIDTH behavior.

## Testing

Ran the full SWF regression suite:

```text
cargo test -p tests --features imgtests --test tests
4372 passed; 0 failed; 21 ignored
```

Also ran the targeted tests for this PR:

```text
cargo test -p tests --features imgtests --test tests avm2/element_format_defaults
cargo test -p tests --features imgtests --test tests avm2/textline_stub_metrics
cargo test -p tests --features imgtests --test tests avm2/textline_fte_basic_metrics
```

## Checklist

- [x] I, a human, have self-reviewed this PR and fully understand the changes within.
- [x] I have [made or updated tests](https://github.com/ruffle-rs/ruffle/blob/master/CONTRIBUTING.md#test-guidelines) where possible.
- [x] All of my commits are properly scoped, compile successfully, and pass all tests.
- [x] This PR does not make sense to split up into smaller PRs.
- [x] An LLM was involved in the authoring of this code.

### CODE REVIEW

File: tests/tests/swfs/avm2/textblock_createline_slots/Test.as [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package {
+import flash.display.Sprite;
+import flash.text.engine.*;
+
+[SWF(width="100", height="100")]
+public class Test extends Sprite {
+    public function Test() {
+        var fd:FontDescription = new FontDescription();
+        fd.fontName = "_sans";
```
Comment (kjarosh) [inline tests/tests/swfs/avm2/textblock_createline_slots/Test.as]: Same issue I flagged in the previous PR.

File: tests/tests/swfs/avm2/textblock_createline_slots/Test.as [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package {
```
Comment (kjarosh) [inline tests/tests/swfs/avm2/textblock_createline_slots/Test.as]: What's the point of this test? It's named "textblock_createline_slots", what does it mean?

File: tests/tests/swfs/avm2/textline_stub_metrics/Test.as [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        var fd:FontDescription = new FontDescription();
+        fd.fontName = "Liberation Sans";
+        fd.fontLookup = FontLookup.DEVICE;
+        var fmt:ElementFormat = new ElementFormat(fd, 14);
+        var narrow:TextLine = new TextBlock(new TextElement("i", fmt))
+            .createTextLine(null, 500);
+        var text:TextLine = new TextBlock(new TextElement("Hello", fmt))
+            .createTextLine(null, 500);
+
+        trace("textWidth follows glyph advance (these are still stubs): " + (text.textWidth > narrow.textWidth));
```
Comment (kjarosh) [inline tests/tests/swfs/avm2/textline_stub_metrics/Test.as]: If you implement them, they will not be stubs, it doesn't matter from the test's perspective

Comment (kjarosh) [inline tests/tests/swfs/avm2/textline_stub_metrics/Test.as]: You can print those values instead of doing math on them

File: tests/tests/swfs/avm2/textline_stub_metrics/Test.as [hunk 4]
```diff
@@ -0,0 +1,25 @@
+package {
+import flash.display.Sprite;
+import flash.text.engine.*;
+
+[SWF(width="100", height="100")]
+public class Test extends Sprite {
+    public function Test() {
+        var fd:FontDescription = new FontDescription();
+        fd.fontName = "Liberation Sans";
+        fd.fontLookup = FontLookup.DEVICE;
```
Comment (kjarosh) [inline tests/tests/swfs/avm2/textline_stub_metrics/Test.as]: Same feedback as in the previous PR: why device fonts?

File: tests/tests/swfs/avm2/textline_stub_metrics/Test.as (hunk 3 above)
Comment (doctorpangloss) [reply tests/tests/swfs/avm2/textline_stub_metrics/Test.as]: yeah this was my biggest uncertainty. so it is okay to commit an xfail with the wrong numbers? they will be fixed "later" and they are stubs in the sense of providing sensible values but they are not the correct values. what is your opinion?

Comment (kjarosh) [reply tests/tests/swfs/avm2/textline_stub_metrics/Test.as]: Committing known failures is fine. Stubs are methods that do not work and return dummy values, buggy methods are not stubs.

File: tests/tests/swfs/avm2/textblock_createline_slots/Test.as (hunk 2 above)
Comment (doctorpangloss) [reply tests/tests/swfs/avm2/textblock_createline_slots/Test.as]: a contrived way to touch some members in textline for the purposes of code coverage. it happens to have a real observable in flash. it fails in master because lastLine is implemented here and not there.

Comment (kjarosh) [reply tests/tests/swfs/avm2/textblock_createline_slots/Test.as]: I see that the only thing you're adding is `textBlockBeginIndex`, which means you should add tests for it specifically. But at the same time it doesn't look like you're implementing it?

I guess I don't understand the scope of the commit/PR.

File: core/src/avm2/globals/flash/text/engine/text_block.rs [hunk 5]
```diff
@@ -12,10 +11,55 @@ use crate::avm2::globals::slots::flash_text_engine_text_line as line_slots;
...
+
+const TEXT_LINE_MAX_LINE_WIDTH: f64 = 1_000_000.0;
+
+fn format_from_content<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    content: Object<'gc>,
+) -> Result<TextFormat, Error<'gc>> {
+    {
+        let context = &mut activation.context;
+        context_stub!(
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/engine/text_block.rs]: Why does this use `context_stub!`?

File: core/src/avm2/globals/flash/text/engine/text_block.rs [hunk 6]
```diff
@@ -27,79 +71,116 @@ pub fn create_text_line<'gc>(
...
-    let content = if matches!(content, Value::Null) {
+    if matches!(content, Value::Null) {
         return Ok(Value::Null);
-    } else {
-        content
+    }
+
+    let text_name = AvmString::new_utf8(activation.gc(), "text");
+    let text = content
+        .get_public_property(text_name, activation)
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/engine/text_block.rs]: Why does this use `get_public_property` instead of `call_method`, like how it was before?

File: core/src/avm2/globals/flash/text/engine/text_block.rs [hunk 7]
```diff
@@ -27,79 +71,116 @@ pub fn create_text_line<'gc>(
...
+        this.set_slot(
+            block_slots::_TEXT_LINE_CREATION_RESULT,
+            istr!("complete").into(),
+            activation,
+        )?;
+        return Ok(Value::Null);
+    }
 
-    let fallback = EditText::new_fte(activation.context, movie.clone(), 0.0, 0.0, width, 15.0);
+    let Some(content_obj) = content.as_object() else {
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/engine/text_block.rs]: The `content` slot can only be either an `Object` or `null`. Why does this check again if it's an object?

File: core/src/avm2/globals/flash/text/engine/text_line.rs [hunk 8]
```diff
@@ -1,16 +1,30 @@
 use crate::avm2::activation::Activation;
 use crate::avm2::error::Error;
 use crate::avm2::value::Value;
+use crate::avm2_stub_getter;
+use crate::display_object::TextLineLayout;
+use std::cell::Ref;
+
+fn text_line_layout<'gc>(this: Value<'gc>) -> Option<Ref<'gc, TextLineLayout<'gc>>> {
+    let line = this.as_object()?.as_display_object()?.as_text_line()?;
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/engine/text_line.rs]: When can any of these ever be `None`? If it's guaranteed, it's fine to use `expect` or even `unwrap`.

File: core/src/avm2/globals/flash/text/engine/text_line.rs [hunk 9]
```diff
@@ -29,10 +43,16 @@ pub fn get_text_height<'gc>(
     this: Value<'gc>,
     _args: &[Value<'gc>],
 ) -> Result<Value<'gc>, Error<'gc>> {
+    if let Some(line) = text_line_layout(this) {
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/engine/text_line.rs]: Doesn't this mean the if-block coming after this one won't ever be hit?

File: core/src/avm2/globals/flash/text/engine/text_line.rs [hunk 10]
```diff
@@ -45,3 +65,29 @@ pub fn get_text_height<'gc>(
...
+pub fn get_ascent<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    this: Value<'gc>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let Some(line) = text_line_layout(this) else {
+        avm2_stub_getter!(activation, "flash.text.engine.TextLine", "ascent");
+        return Ok(12.0.into());
+    };
+    avm2_stub_getter!(activation, "flash.text.engine.TextLine", "ascent");
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/engine/text_line.rs]: This code should be unreachable as the display object of a `TextLine` is always a `TextLine`, right?

File: core/src/avm2/globals/flash/text/engine/text_line.rs [hunk 11]
```diff
@@ -45,3 +65,29 @@ pub fn get_text_height<'gc>(
...
+pub fn get_descent<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    this: Value<'gc>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let Some(line) = text_line_layout(this) else {
+        avm2_stub_getter!(activation, "flash.text.engine.TextLine", "descent");
+        return Ok(3.0.into());
+    };
+    avm2_stub_getter!(activation, "flash.text.engine.TextLine", "descent");
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/engine/text_line.rs]: Same as above

File: core/src/avm2/globals/flash/text/engine/text_line.rs [hunk 12]
```diff
@@ -45,3 +65,29 @@ pub fn get_text_height<'gc>(
     let measured_text = edit_text.measure_text(activation.context);
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/engine/text_line.rs]: Related to changes made in your previous PR- when can `display_object.as_edit_text()` ever be `Some`? Is it ever possible for a `TextLine` to have a non-`TextLine` display object?

File: core/src/display_object/text_line.rs [hunk 13]
```diff
@@ -44,19 +89,25 @@ impl<'gc> TextLine<'gc> {
...
+        line: TextLineLayout<'gc>,
         fallback: Option<EditText<'gc>>,
     ) -> Self {
         TextLine(Gc::new(
             context.gc(),
             TextLineData {
                 base: Default::default(),
                 avm2_object: Lock::new(None),
+                line: RefLock::new(line),
                 fallback,
```
Comment (Lord-McSweeney) [inline core/src/display_object/text_line.rs:94]: When is `fallback` ever `None`?

File: core/src/avm2/globals/flash/text/engine/text_block.rs [hunk 14]
```diff
@@ -27,79 +64,115 @@ pub fn create_text_line<'gc>(
...
+    }
 
+    let content_obj = content
+        .as_object()
+        .expect("TextBlock content slot must be null or ContentElement");
+    let spans = FormatSpans::from_text(
+        WString::from(text.as_wstr()),
+        format_from_content(activation, content_obj)?,
+    );
+    let requested_width = if args.get_f64(1) >= TEXT_LINE_MAX_LINE_WIDTH {
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/engine/text_block.rs]: `args.get_f64(1)` is called three times; can you keep it as a variable like how it was before?

File: core/src/avm2/globals/flash/text/engine/text_block.rs [hunk 15]
```diff
@@ -13,9 +13,46 @@ use crate::avm2::object::{Object, TObject};
...
+        font: Some(WString::from_utf8("_serif")),
+        size: Some(12.0),
+        color: Some(swf::Color::from_rgb(0, 0xff)),
+        ..Default::default()
+    };
+
+    if let Some(ef) = content.get_slot(element_slots::_ELEMENT_FORMAT).as_object() {
+        let color = ef
+            .get_slot(format_slots::_COLOR)
+            .coerce_to_u32(activation)?;
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/engine/text_block.rs]: This can be `as_u32`

File: core/src/avm2/globals/flash/text/engine/text_block.rs [hunk 16]
```diff
@@ -13,9 +13,46 @@ use crate::avm2::object::{Object, TObject};
...
+        format.color = Some(swf::Color::from_rgb(color & 0xff_ffff, 0xff));
+        format.size = Some(
+            ef.get_slot(format_slots::_FONT_SIZE)
+                .coerce_to_number(activation)?,
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/engine/text_block.rs]: This can be `as_f64`

File: tests/tests/swfs/avm2/textblock_createline_slots/Test.as (hunk 2 above)
Comment (doctorpangloss) [reply tests/tests/swfs/avm2/textblock_createline_slots/Test.as]: the test needed to be bigger and now it is. more stuff correctly fails on master and passes here too

File: core/src/avm2/globals/flash/text/engine/text_block.rs [hunk 17]
```diff
@@ -31,75 +63,105 @@ pub fn create_text_line<'gc>(
...
+        &spans,
+        activation.context,
+        movie.clone(),
+        requested_width,
+        false,
+        true,
+        FontType::Device,
+    );
+    let Some(html_line) = layout.lines().first().cloned() else {
+        return Ok(Value::Null);
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/engine/text_block.rs]: Could you add test coverage for this case? Right now it seems that none of the tests hit this

File: tests/tests/swfs/avm2/textline_fte_basic_metrics/Test.as [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+            .createTextLine(null, 100000);
+        var wrappedText:String = "one two three";
+        var narrowBlock:TextBlock = new TextBlock(new TextElement(wrappedText, new ElementFormat(fd, 30)));
+        var narrowLine:TextLine = narrowBlock.createTextLine(null, 60);
+        var maxLine:TextLine = new TextBlock(new TextElement(wrappedText, new ElementFormat(fd, 30)))
+            .createTextLine(null, TextLine.MAX_LINE_WIDTH);
+        var aboveMaxLine:TextLine = new TextBlock(new TextElement(wrappedText, new ElementFormat(fd, 30)))
+            .createTextLine(null, TextLine.MAX_LINE_WIDTH + 1);
+
+        trace("line.textWidth: " + rounded(line.textWidth));
```
Comment (Lord-McSweeney) [inline tests/tests/swfs/avm2/textline_fte_basic_metrics/Test.as]: Can you test all of these properties (`textWidth`, `textHeight`, `ascent`, `descent`, `rawTextLength`) on all of the created lines?

Also, is there a reason this doesn't also test `line.width`?

File: core/src/avm2/globals/flash/text/engine/text_block.rs [hunk 19]
```diff
@@ -119,9 +119,11 @@ pub fn create_text_line<'gc>(
...
         FontType::Device,
     );
-    let Some(html_line) = layout.lines().first().cloned() else {
-        return Ok(Value::Null);
-    };
+    let html_line = layout
+        .lines()
+        .first()
+        .cloned()
+        .expect("TextLine layout must contain a line for nonempty text");
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/engine/text_block.rs:106]: What happens if the text is actually empty?

Comment (doctorpangloss) [reply core/src/avm2/globals/flash/text/engine/text_block.rs:106]: a new xfail test is in, and addressing it will come later when more machinery is implemented to do that

**Comment by doctorpangloss** [comment]
there is another xfail case that covers blank and whitespace in textlines

File: tests/tests/swfs/fte/element_format_defaults/Test.as [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+    public class Test extends Sprite {
+        public function Test() {
+            var format:ElementFormat = new ElementFormat();
+
+            trace("fontDescription.fontName: " + format.fontDescription.fontName);
+            trace("fontDescription.fontWeight: " + format.fontDescription.fontWeight);
+            trace("fontDescription.fontPosture: " + format.fontDescription.fontPosture);
+            trace("fontDescription.fontLookup: " + format.fontDescription.fontLookup);
+            trace("fontDescription.renderingMode: " + format.fontDescription.renderingMode);
+            trace("fontDescription.cffHinting: " + format.fontDescription.cffHinting);
```
Comment (kjarosh) [inline tests/tests/swfs/fte/element_format_defaults/Test.as:14]: Missing `fontDescription.locked`.

File: tests/tests/swfs/fte/element_format_defaults/Test.as [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+            trace("baselineShift: " + format.baselineShift);
+            trace("kerning: " + format.kerning);
+            trace("trackingRight: " + format.trackingRight);
+            trace("trackingLeft: " + format.trackingLeft);
+            trace("locale: " + format.locale);
+            trace("breakOpportunity: " + format.breakOpportunity);
+            trace("digitCase: " + format.digitCase);
+            trace("digitWidth: " + format.digitWidth);
+            trace("ligatureLevel: " + format.ligatureLevel);
+            trace("typographicCase: " + format.typographicCase);
```
Comment (kjarosh) [inline tests/tests/swfs/fte/element_format_defaults/Test.as:31]: Missing `locked`.

File: tests/tests/swfs/avm2/textblock_fte_empty_content/Test.as [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+            trace(name + ".textWidth: " + rounded(line.textWidth));
+            trace(name + ".textHeight: " + rounded(line.textHeight));
+            trace(name + ".ascent: " + rounded(line.ascent));
+            trace(name + ".descent: " + rounded(line.descent));
+            trace(name + ".rawTextLength: " + line.rawTextLength);
+        }
+    }
+
+    private function rounded(value:Number):Number {
+        return Math.round(value * 100) / 100;
```
Comment (kjarosh) [inline tests/tests/swfs/avm2/textblock_fte_empty_content/Test.as]: Why rounding?

File: tests/tests/swfs/avm2/textblock_empty_content_basic/Test.as [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+import flash.display.Sprite;
+import flash.text.engine.ElementFormat;
+import flash.text.engine.FontDescription;
+import flash.text.engine.FontLookup;
+import flash.text.engine.TextBlock;
+import flash.text.engine.TextElement;
+import flash.text.engine.TextLine;
+
+[SWF(width="200", height="100")]
+public class Test extends Sprite {
```
Comment (kjarosh) [inline tests/tests/swfs/avm2/textblock_empty_content_basic/Test.as]: What's the point of this test compared to `avm2/textblock_fte_empty_content`?

File: tests/tests/swfs/fte/textblock_empty_content/Test.as [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+    public function Test() {
+        var fd:FontDescription = new FontDescription();
+        fd.fontName = "Liberation Sans";
+        fd.fontLookup = FontLookup.EMBEDDED_CFF;
+        var fmt:ElementFormat = new ElementFormat(fd, 20);
+
+        traceCase("empty", "", fmt);
+        traceCase("space", " ", fmt);
+        traceCase("twoSpaces", "  ", fmt);
+        traceCase("newline", "\n", fmt);
```
Comment (kjarosh) [inline tests/tests/swfs/fte/textblock_empty_content/Test.as:24]: You might want to also cover `\r` and `\r\n`.

File: tests/tests/swfs/fte/textblock_empty_content/Test.as [hunk 25]
```diff
@@ -0,0 +1,47 @@
...
+
+        trace(name + ".line is null: " + (line === null));
+        trace(name + ".creation result: " + block.textLineCreationResult);
+        if (line !== null) {
+            trace(name + ".width: " + rounded(line.width));
+            trace(name + ".textWidth: " + rounded(line.textWidth));
+            trace(name + ".textHeight: " + rounded(line.textHeight));
+            trace(name + ".ascent: " + rounded(line.ascent));
+            trace(name + ".descent: " + rounded(line.descent));
+            trace(name + ".rawTextLength: " + line.rawTextLength);
```
Comment (kjarosh) [inline tests/tests/swfs/fte/textblock_empty_content/Test.as:46]: Can you add the rest of properties?

File: tests/tests/swfs/fte/textblock_empty_content/Test.as
Comment (kjarosh) [inline tests/tests/swfs/fte/textblock_empty_content/Test.as:1]: Nit: you don't need to include `fte` in the name, textblock is a part of FTE so it's obvious.

File: tests/tests/swfs/avm2/textblock_empty_content_basic/Test.as (hunk 23 above)
Comment (doctorpangloss) [reply tests/tests/swfs/avm2/textblock_empty_content_basic/Test.as]: it shows what is correctly implemented in this commit, it passes, as opposed to the other stuff which xfails

**Comment by doctorpangloss** [comment]
this one is ready

File: tests/tests/swfs/avm2/textblock_empty_content_basic/Test.as (hunk 23 above)
Comment (Lord-McSweeney) [reply tests/tests/swfs/avm2/textblock_empty_content_basic/Test.as]: Is there anything covered in `textblock_empty_content_basic` that's not covered in `textblock_empty_content`? We don't count which of our tests are `known_failure`, and we do have expected output for each of those tests, so a `known_failure` test is just as useful as a passing test. If this test is redundant it's best to remove it.

**Comment by Lord-McSweeney** [comment]
Mentioning "textline" and "textblock" in a test name that already mentions "fte" is redundant. For those tests the "fte" in their names can be removed.

In the future, could you move all the fte tests to a folder in `tests/tests/swfs`, such as `tests/tests/swfs/fte/`?

File: tests/tests/swfs/fte/textblock_empty_content/Test.as [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+        fd.fontName = "Liberation Sans";
+        fd.fontLookup = FontLookup.EMBEDDED_CFF;
+        var fmt:ElementFormat = new ElementFormat(fd, 20);
+
+        traceCase("empty", "", fmt);
+        traceCase("space", " ", fmt);
+        traceCase("twoSpaces", "  ", fmt);
+        traceCase("newline", "\n", fmt);
+        traceCase("carriageReturn", "\r", fmt);
+        traceCase("crlf", "\r\n", fmt);
```
Comment (Lord-McSweeney) [inline tests/tests/swfs/fte/textblock_empty_content/Test.as:26]: Could you also add `"\t"` and `" \n "`?

File: core/src/avm2/globals/flash/text/engine/TextLine.as [hunk 27]
```diff
@@ -25,6 +25,15 @@ package flash.text.engine {
         [Ruffle(NativeAccessible)]
         private var _rawTextLength:int = 0;
 
+        [Ruffle(NativeAccessible)]
+        private var _textBlockBeginIndex:int = 0;
+
+        [Ruffle(NativeAccessible)]
+        internal var _nextLine:TextLine = null;
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/engine/TextLine.as]: This can be `private` instead of `internal`

File: core/src/avm2/globals/flash/text/engine/TextLine.as [hunk 28]
```diff
@@ -25,6 +25,15 @@ package flash.text.engine {
...
+
+        [Ruffle(NativeAccessible)]
+        internal var _previousLine:TextLine = null;
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/engine/TextLine.as]: This can be `private` instead of `internal`

File: core/src/avm2/globals/flash/text/engine/TextLine.as [hunk 29]
```diff
@@ -25,6 +25,15 @@ package flash.text.engine {
         [Ruffle(NativeAccessible)]
         private var _rawTextLength:int = 0;
 
+        [Ruffle(NativeAccessible)]
+        private var _textBlockBeginIndex:int = 0;
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/engine/TextLine.as:29]: For the future- it'd be best to have all these properties be fields on `TextLineData`, instead of Actionscript fields.

File: core/src/avm2/globals/flash/text/engine/text_block.rs [hunk 30]
```diff
@@ -13,9 +13,41 @@ use crate::avm2::object::{Object, TObject};
...
-use crate::string::WStr;
+use crate::display_object::{EditText, TDisplayObject, TextLine, TextLineLayout};
+use crate::font::FontType;
+use crate::html::{FormatSpans, TextFormat, lower_from_text_spans};
+use crate::string::{WStr, WString};
+use swf::Twips;
+
+const TEXT_LINE_MAX_LINE_WIDTH: f64 = 1_000_000.0;
+
+fn format_from_content<'gc>(
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/engine/text_block.rs]: Isn't this whole function redundant with `apply_format`? If the last few lines of `apply_format` were removed, couldn't it be used in place of this function?

File: tests/tests/swfs/avm2/textblock_empty_content/output.ruffle.txt [hunk 31]
```diff
@@ -0,0 +1,107 @@
+empty.line is null: true
+empty.creation result: complete
+space.line is null: false
+space.creation result: success
+space.width: 100
+space.textWidth: 6
+space.textHeight: 22.299999237060547
+space.ascent: 18.100000381469727
+space.descent: 4.199999809265137
```
Comment (Lord-McSweeney) [inline tests/tests/swfs/avm2/textblock_empty_content/output.ruffle.txt]: The `textHeight`, `ascent`, and `descent` info look totally wrong (comparing this output from Ruffle to the output from FP), are you sure the calculation logic for these is implemented correctly? Could you mark the AS methods to access them as stubs?

File: core/src/avm2/globals/flash/text/engine/ElementFormat.as [hunk 32]
```diff
@@ -21,6 +23,7 @@ package flash.text.engine {
         private var _kerning:String;
         private var _ligatureLevel:String;
         private var _locale:String;
+        private var _locked:Boolean = false;
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/engine/ElementFormat.as]: nit: you can remove the ` = false` here or move it to the constructor, like how the rest of the properties work

File: core/src/display_object/text_line.rs [hunk 33]
```diff
@@ -8,15 +8,59 @@ use crate::display_object::{
...
+#[derive(Collect)]
+#[collect(no_drop)]
+pub struct TextLineLayout<'gc> {
+    html_line: LayoutLine<'gc>,
+    #[collect(require_static)]
+    text: WString,
+    #[collect(require_static)]
+    ascent: f32,
+    #[collect(require_static)]
+    descent: f32,
```
Comment (Lord-McSweeney) [inline core/src/display_object/text_line.rs]: Why are these stored as `f32` and not as `Twips`?

File: core/src/display_object/text_line.rs [hunk 34]
```diff
@@ -8,15 +8,59 @@ use crate::display_object::{
...
+            descent,
+        }
+    }
+
+    pub fn ascent(&self) -> f32 {
+        self.ascent
+    }
+
+    pub fn descent(&self) -> f32 {
+        self.descent
```
Comment (Lord-McSweeney) [inline core/src/display_object/text_line.rs]: Why are these stored as separate properties? Is it that expensive to calculate them from `html_line.descent()`?

File: core/src/avm2/globals/flash/text/engine/TextLine.as [hunk 35]
```diff
@@ -75,8 +91,15 @@ package flash.text.engine {
...
-        public native function get textHeight():Number;
+        public function get textWidth():Number {
+            stub_getter("flash.text.engine.TextLine", "textWidth");
+            return this._rawTextLength * 6;
+        }
+
+        public function get textHeight():Number {
+            stub_getter("flash.text.engine.TextLine", "textHeight");
+            return 22.3;
+        }
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/engine/TextLine.as]: Why were these two made into stubs?

File: core/src/avm2/globals/flash/text/engine/TextLine.as [hunk 36]
```diff
@@ -48,9 +56,17 @@ package flash.text.engine {
...
+
+        public function get previousLine():TextLine {
+            return this._previousLine;
+        }
+
         public function get ascent():Number {
             stub_getter("flash.text.engine.TextLine", "ascent");
-            return 12.0;
+            return 18.1;
```
Comment (kjarosh) [inline core/src/avm2/globals/flash/text/engine/TextLine.as:69]: Why this change?

File: core/src/avm2/globals/flash/text/engine/text_block.rs [hunk 37]
```diff
@@ -109,69 +146,71 @@ fn apply_format<'gc>(
     display_object: EditText<'gc>,
     text: &WStr,
     element_format: Option<Object<'gc>>,
-) -> Result<(), Error<'gc>> {
-    if let Some(element_format) = element_format {
-        // TODO: Support more ElementFormat properties
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/text/engine/text_block.rs:114]: Does the new implementation support all the ElementFormat properties? If not, could you keep this TODO comment in?
